### PR TITLE
Resolve symlinks in cp

### PR DESCRIPTION
**Pull request by mheon**
Fixes CVE-2018-15664 and #3211 

Still needs manpage updates, bash completion updates, and a test

### CODE REVIEW

**Comment by vrothberg** [comment]
That looks very painless :+1: Changed LGTM but I'd love to have some tests for #3211.

**Comment by QiWang19** [comment]
LGTM

**Comment by baude** [comment]
lgtm, test would be nth

**Comment by mheon** [comment]
Added a test

**Comment by mheon** [comment]
Update: fix here is still vulnerable to races. I need to move copy code inside libpod to avoid them (so we can grab the container lock while copying)

**Comment by mheon** [comment]
Alright, I think this is "good enough" for a release. There's further work to be done, but that will require a much larger refactor.

**Comment by mheon** [comment]
Hm. Rootless containers don't have CGroups, so we can't pause them.

As such, we can't really mitigate for them... Hm.

**Comment by mheon** [comment]
With any luck, tests will go green now.

**Comment by rhatdan** [comment]
@mheon How about we fail and say you can not copy on a running container when in rootless mode?

**Comment by mheon** [comment]
Good idea - pushed another commit. We now error on trying to copy into running rootless container with `--pause` set (which is the default).

You can still do this with `--pause=false`

**Comment by mheon** [comment]
OK, tests mostly passing. Restarted one flake.
@baude @rhatdan Someone want to LGTM again?

**Comment by sfowl** [comment]
CVE-2019-10152 has been assigned for this issue:

https://bugzilla.redhat.com/show_bug.cgi?id=1715667

**Comment by ret2libc** [comment]
What about the problems outlined in https://github.com/moby/moby/pull/39252#issuecomment-497261631 ? It seems docker is not going to pause containers, for what I understand, because it is a breaking change (e.g. copying big files may stop the process for too long and breaking network connections).

Is podman affected by the same issue?

File: cmd/podman/cp.go [hunk 1]
```diff
@@ -93,6 +95,38 @@ func copyBetweenHostAndContainer(runtime *libpod.Runtime, src string, dest strin
...
+			// TODO: We can potentially start the container while
+			// the copy is running, which still allows a race where
+			// malicious code could mess with the symlink.
+			if errors.Cause(err) != libpod.ErrCtrStateInvalid {
+				return err
+			}
+		} else if err == nil {
+			// Only add the defer if we actually paused
+			defer func() {
+				if err := ctr.Unpause(); err != nil {
```
Comment (giuseppe) [inline cmd/podman/cp.go]: is it possible that two processes are cp'ing at the same time and one of them unpause the container while the other process is still resolving the symlinks?

Comment (mheon) [reply cmd/podman/cp.go]: Yep. That's part of why I was looking to move `cp` into the container.

Comment (mheon) [reply cmd/podman/cp.go]: Into libpod, rather

**Comment by mheon** [comment]
@ret2libc Hm. We should be able to grab that - it's definitely a more elegant solution than this. We'll keep this in for now, but I'll avoid cutting a release with it until we verify if we can swap to the alternative fix - in which case we can revert this.

**Comment by rhatdan** [comment]
Yes I like the Moby Solution better, and we need to back port that ASAP as soon as it is merged.

**Pull request merged**